lib/src/group_changes.dart

```diff
@@ -0,0 +1,307 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Add fields with data about the test run and the commit tested, and
+// with the result on the last build tested, to the test results file.
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:resource/resource.dart' show Resource;
+
+import 'package:dart_ci/src/fetch_changes.dart';
+
+class MinMax {
+  int min = null;
+  int max = null;
+
+  void add(int value) {
+    if (value == null) return;
+    if (min == null || value < min) min = value;
+    if (max == null || value > max) max = value;
+  }
+
+  String toString() => "($min:$max)";
+}
+
+Map<String, List<String>> configSets;
+int idCounter = 1;
+
+class ByConfigSetData {
+  String formattedConfigSet;
+  String test; // Test name, including suite
+  String change; // Change in results, reported as string "was, now, expected"
+  MinMax before; // MinMax of indices of commits before this change
+  MinMax after; // MinMax of indices of commits after or at this change.
+
+  ByConfigSetData(
+      this.formattedConfigSet, this.test, this.change, this.before, this.after);
+}
+
+Future<String> createChangesPage() async {
+  var hashes = <String>[];
+  var commitData = <String>[];
+  var reviewLinks = <String>[];
+  const reviewPrefix = "Reviewed-on: ";
+  // Load data from gerrit/git REST API
+  final commits = (await commitInformation())["log"] as List<dynamic>;
+  for (Map<String, dynamic> commit in commits) {
+    hashes.add(commit["commit"]);
+    reviewLinks.add(commit["message"]
+        .split('\n')
+        .firstWhere((String line) => line.startsWith(reviewPrefix),
+            orElse: () =>
+                "${reviewPrefix}https://dart.googlesource.com/sdk/+/${hashes.last}")
+        .substring(reviewPrefix.length));
+    final time = commit["committer"]["time"].substring(0, 16);
+    final author = commit["author"]["email"];
+    final subject = commit["message"].split('\n').first;
+    commitData.add("$time $author $subject");
+  }
+
+  // Changes come from global server data, fetched by server.
+  final data = computePageData(changes, hashes);
+  return htmlPage(data, hashes, commitData, reviewLinks);
+}
+
+Future<Map<String, dynamic>> commitInformation() async {
+  final client = HttpClient();
+  final request = await client.getUrl(Uri.parse(
+      "https://dart.googlesource.com/sdk/+log/master?n=400&format=JSON"));
```

Comment:
Consider getting the entire history instead of just 400 and making the order chronological rather than reverse chronological. Subsequent requests could do an incremental update of the history to be cheaper on GoB. Although if we are to have stable commit numbering as a key, that would require handling merges, not sure that's possible here.